runtime/compiler/optimizer/JProfilingValue.cpp

```diff
@@ -491,171 +488,127 @@ TR_JProfilingValue::addProfilingTrees(
+   TR::Block *quickTest = NULL;
    TR::SymbolReference *storedValueSymRef = NULL;
-   TR::TreeTop *storeValue = TR::TreeTop::create(comp, insertionPoint, storeNode(comp, value, storedValueSymRef));
-   if (!optionalTest)
-      replaceNode(comp, extendedBlock->getEntry(), storeValue->getNextTreeTop(),
-         value, TR::Node::createLoad(example, storeValue->getNode()->getSymbolReference()));
-
-   // Split after the store
-   TR::Block *quickTest = originalBlock->split(storeValue->getNextTreeTop(), cfg, true, true);
-
-   /********************* fallback Block *********************/
-
-   // Insert the optional test and split after it
+   TR::TreeTop *quickTestLastProfilingTT = NULL;
+   
+   // And to avoid NPE in profiling trees we need to check if that object whose J9Class we are going to load is not NULL
    if (optionalTest)
       {
-      TR_ASSERT_FATAL(fallbackValue, "If introducing an optional test, a fallback value must be specified");
-      TR::TreeTop *testTree = TR::TreeTop::create(comp, insertionPoint, optionalTest);
-
-      TR::Block *mapping = originalBlock->split(testTree->getNextTreeTop(), cfg, true, true);
-      if (extendBlocks)
-         mapping->setIsExtensionOfPreviousBlock();
-      
-      // Create the fallback block and link it to the optional test
-      TR::Block *fallback = TR::Block::createEmptyBlock(comp, MAX_COLD_BLOCK_COUNT + 1);
-      fallback->setIsCold();
-      lastTreeTop->join(fallback->getEntry());
-      lastTreeTop = fallback->getExit();
-      cfg->addNode(fallback);
-      cfg->addEdge(originalBlock, fallback);
-      cfg->addEdge(fallback, quickTest);
-      testTree->getNode()->setBranchDestination(fallback->getEntry());     
-
-      if (trace)
-         traceMsg(comp, "  Fallback block_%d store to symRef #%d of node %p in \n", fallback->getNumber(),
-            storedValueSymRef->getReferenceNumber(), fallback);
-
-      // Store and return to the mainline
-      TR::TreeTop::create(comp, fallback->getEntry(), TR::Node::create(example, TR::Goto, 0, quickTest->getEntry()));
-      TR::TreeTop::create(comp, fallback->getEntry(), storeNode(comp, fallbackValue, storedValueSymRef));
+      TR::SymbolReference *storedObjectSymRef = NULL;
+      // We add profiling trees before checkCast or instanceOf node and to avoid commoning out object while splitting blocks,
+      // object is stored in temp slot and subbsequent reference to that object is replaced by loading from that temp slot
+      TR::TreeTop *storeValue = TR::TreeTop::create(comp, insertionPoint, storeNode(comp, value->getFirstChild(), storedObjectSymRef));
+      replaceNode(comp, extendedBlock->getEntry(), storeValue->getNextTreeTop(),
+         value->getFirstChild(), TR::Node::createLoad(value->getFirstChild(), storedObjectSymRef));
+      // VFT pointer is required to stored in temp slot for helper call.
+      quickTest = originalBlock->split(storeValue->getNextTreeTop(), cfg, true, true);
```

Comment:
We may not always be profiling a vft pointer - will this logic always be right? I suspect what you mean is that the value to be profiled needs to be stored to a temp so it can be read in the helper call block, but I wanted to check if there was something vft specific